File: server/src/routes/reviews.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { prisma } from '../db';
import { authRequired } from '../middleware/auth';

const router = Router();

const reviewSchema = z.object({
  rating: z.number().min(1).max(5),
  content: z.string().max(2000).default(''),
  images: z.array(z.string()).default([]),
});

router.get('/location/:locationId', async (req: Request, res: Response) => {
  const locationId = parseInt(req.params.locationId);
  const reviews = await prisma.review.findMany({
    where: { locationId },
    include: { user: { select: { id: true, username: true, avatar: true } } },
    orderBy: { createdAt: 'desc' }
  });
  const parsed = reviews.map(r => ({ ...r, images: JSON.parse(r.images) }));
  res.json(parsed);
});

router.post('/:locationId', authRequired, async (req: Request, res: Response) => {
  try {
    const locationId = parseInt(req.params.locationId);
    const data = reviewSchema.parse(req.body);
    const existing = await prisma.review.findFirst({
      where: { userId: req.user!.userId, locationId }
    });
    if (existing) {
      const review = await prisma.review.update({
        where: { id: existing.id },
        data: { rating: data.rating, content: data.content, images: JSON.stringify(data.images) }
      });
      return res.json(review);
    }
    const review = await prisma.review.create({
      data: {
        userId: req.user!.userId,
        locationId,
        rating: data.rating,
        content: data.content,
        images: JSON.stringify(data.images)
      }
    });
    res.json(review);
  } catch (e: any) {
    if (e instanceof z.ZodError) return res.status(400).json({ error: '数据格式错误' });
    res.status(500).json({ error: '提交评价失败' });
  }
});

router.delete('/:id', authRequired, async (req: Request, res: Response) => {
  const id = parseInt(req.params.id);
  const review = await prisma.review.findUnique({ where: { id } });
  if (!review) return res.status(404).json({ error: '评价不存在' });
  if (review.userId !== req.user!.userId && req.user!.role !== 'admin') {
    return res.status(403).json({ error: '无权限' });
  }
  await prisma.review.delete({ where: { id } });
  res.json({ success: true });
});

export default router;

File: server/routes/reviews.py
import json
from flask import Blueprint, request, jsonify
from models import db, Review
from utils.auth import auth_required

reviews_bp = Blueprint('reviews', __name__)


@reviews_bp.route('/location/<int:location_id>', methods=['GET'])
def list_reviews(location_id):
    reviews = Review.query.filter_by(location_id=location_id).order_by(Review.created_at.desc()).all()
    return jsonify([r.to_dict() for r in reviews])


@reviews_bp.route('/<int:location_id>', methods=['POST'])
@auth_required
def submit_review(location_id):
    data = request.get_json() or {}
    rating = data.get('rating', 0)
    content = data.get('content', '')
    images = data.get('images', [])
    if not 1 <= rating <= 5:
        return jsonify({'error': '评分需在1-5之间'}), 400
    if len(content) > 2000:
        return jsonify({'error': '内容不能超过2000字'}), 400
    existing = Review.query.filter_by(user_id=request.user_id, location_id=location_id).first()
    if existing:
        existing.rating = rating
        existing.content = content
        existing.images = json.dumps(images, ensure_ascii=False)
        db.session.commit()
        return jsonify(existing.to_dict())
    review = Review(
        user_id=request.user_id,
        location_id=location_id,
        rating=rating,
        content=content,
        images=json.dumps(images, ensure_ascii=False),
    )
    db.session.add(review)
    db.session.commit()
    return jsonify(review.to_dict()), 201


@reviews_bp.route('/<int:id>', methods=['DELETE'])
@auth_required
def delete_review(id):
    review = db.session.get(Review, id)
    if not review:
        return jsonify({'error': '评价不存在'}), 404
    if review.user_id != request.user_id and getattr(request, 'user_role', None) != 'admin':
        return jsonify({'error': '无权限'}), 403
    db.session.delete(review)
    db.session.commit()
    return jsonify({'success': True})
